src/corecel/cont/Bitset.hh

```diff
@@ -0,0 +1,395 @@
+    {
+        for (size_type i = 0; i < num_words - 1; ++i)
+        {
+            if (words_[i] != ~static_cast<word_type>(0))
+            {
+                return false;
+            }
+        }
+        // only compare the last word up to the last bit of the bitset
+        return this->last_word()
+               == (~static_cast<word_type>(0)
+                   >> (num_words * bits_per_word - N));
+    }
+
+    //! Check if any bits are set
+    CELER_CONSTEXPR_FUNCTION bool any() const noexcept
+    {
+        for (size_type i = 0; i < num_words; ++i)
+        {
+            if (words_[i] != static_cast<word_type>(0))
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    //! Check if no bits are set
+    CELER_CONSTEXPR_FUNCTION bool none() const noexcept
+    {
+        return !this->any();
+    }
+
+    //! Number of bits set to true
+    CELER_CONSTEXPR_FUNCTION size_type count() const noexcept
+    {
+        size_type count = 0;
+        for (size_type i = 0; i < num_words; ++i)
+        {
+            count += celeritas::popcount(words_[i]);
+        }
+
+        return count;
+    }
+
+    //! Number of bits in the bitset
+    CELER_CONSTEXPR_FUNCTION size_type size() const noexcept { return N; }
+
+    //! Bitwise AND with another bitset
+    CELER_CONSTEXPR_FUNCTION Bitset& operator&=(Bitset const& other) noexcept
+    {
+        for (size_type i = 0; i < num_words; ++i)
+        {
+            words_[i] &= other.words_[i];
+        }
+        return *this;
+    }
+
+    //! Bitwise OR with another bitset
+    CELER_CONSTEXPR_FUNCTION Bitset& operator|=(Bitset const& other) noexcept
+    {
+        for (size_type i = 0; i < num_words; ++i)
+        {
+            words_[i] |= other.words_[i];
+        }
+        return *this;
+    }
+
+    //! Bitwise XOR with another bitset
+    CELER_CONSTEXPR_FUNCTION Bitset& operator^=(Bitset const& other) noexcept
+    {
+        for (size_type i = 0; i < num_words; ++i)
+        {
+            words_[i] ^= other.words_[i];
+        }
+        return *this;
+    }
+
+    //! Return a copy with all bits flipped (bianry NOT)
+    CELER_CONSTEXPR_FUNCTION Bitset operator~() const noexcept
+    {
+        return Bitset{*this}.flip();
+    }
+
+    //! Set all bits
+    CELER_CONSTEXPR_FUNCTION Bitset& set() noexcept
+    {
+        for (size_type i = 0; i < num_words; ++i)
+        {
+            words_[i] = ~static_cast<word_type>(0);
+        }
+        // sanitize the last word
+        detail::Sanitize<N % bits_per_word>::sanitize(this->last_word());
+        return *this;
+    }
+
+    //! Set the bit at position pos
+    CELER_CONSTEXPR_FUNCTION Bitset&
+    set(size_type pos, bool value = true) noexcept
+    {
+        if (value)
+        {
+            this->get_word(pos) |= Bitset::mask(pos);
+        }
+        else
+        {
+            this->get_word(pos) &= ~Bitset::mask(pos);
+        }
```

Comment:
Is this any more efficient than `(*this)[pos] = value;`?

Reply:
it _should_ be the same if the compiler can inline everything which seems very easy to do.